message-feed: Fixed narrow banner on All messages screen

SahilSingh177 opened the pull request:
Made changes in reset_ui_state to check if message list is empty or not and on the basis of this narror banner will be shown. Changed narrowed value to true in message list, so that All messages acts like other streams.

Fixes #23701

**Screenshots and screen captures:**

Original Behavior

![204607852-6189b30a-e252-48ef-8766-e299b1bbde1e](https://user-images.githubusercontent.com/96344003/209562194-5ec32b4c-66f6-4f59-b087-86fc029eb8e3.png)


New Behavior after the changes

![Screenshot_20230103_002931](https://user-images.githubusercontent.com/96344003/210270069-bad155c7-f192-43cb-98bb-3f765c3f8d3f.png)



Unlike older tables, home was not acting in the same way during rerendering. So, I made changes in reset_ui_state method, so that in case of no message narrow banner appears.

<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [X] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [X] Explains differences from previous plans (e.g., issue description).
- [X] Highlights technical choices and bugs encountered.
- [X] Calls out remaining decisions and concerns.
- [X] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [X] Each commit is a coherent idea.
- [X] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [X] Visual appearance of the changes.
- [X] Responsiveness and internationalization.
- [X] Strings and tooltips.
- [X] End-to-end functionality of buttons, interactions and flows.
- [X] Corner cases, error conditions, and easily imagined bugs.
</details>

SahilSingh177 commented:
@alya please review this PR. Thanks!

N-Shar-ma commented on static/js/message_list.js:
```diff
@@ -26,7 +26,7 @@ export class MessageList {
         this.view = new MessageListView(this, table_name, collapse_messages);
         this.table_name = table_name;
-        this.narrowed = this.table_name === "zfilt";
+        this.narrowed = true; // Making narrowed true for all tables including home
```
While this seems to work for the use case of the banner, the "all messages" view isn't a narrow so this should not be changed - this might end up affecting more things than intended. It might be better to just not check for this variable when displaying / hiding the banner

SahilSingh177 replied:
> While this seems to work for the use case of the banner, the "all messages" view isn't a narrow so this should not be changed - this might end up affecting more things than intended. It might be better to just not check for this variable when displaying / hiding the banner

Changing narrowed true just for "zfilt" results in the following bug:

https://user-images.githubusercontent.com/96344003/210390678-87bbc4eb-b89a-4a65-9f95-9aba0d011054.mov

Thus, to re-render all messages any time a stream is unmuted, I changed narrowed value to true.

SahilSingh177 replied:
Ok. I'll make the changes.

SahilSingh177 commented:
@N-Shar-ma Please review the PR and suggest if any change is required.

N-Shar-ma commented on static/js/stream_muting.js:
```diff
@@ -21,7 +21,7 @@ export function update_is_muted(sub, value) {
         } else if (
             message_lists.home === message_lists.current &&
-            message_lists.current.selected_row().offset() !== null
+            message_lists.current.selected_row().offset() !== undefined
```
Why is this change needed?

SahilSingh177 replied:
Actually, It was causing an error saying cannot read properties of undefined. I will change this to original code.

N-Shar-ma commented on static/js/message_list.js:
```diff
@@ -68,14 +68,14 @@ export class MessageList {
         }
 
-        if (this.narrowed && !this.empty() && this.selected_id() === -1) {
+        if (!this.empty() && this.selected_id() === -1) {
```
I don't think this line of change is needed since this doesn't affect the narrow_banner

N-Shar-ma reviewed (changes requested):
Have left a couple comments... Also could you fix the commit history? Look at https://zulip.readthedocs.io/en/latest/git/using.html#examine-and-tidy-your-commit-history for guidelines. Note that we don't use merge commits

SahilSingh177 commented:
@N-Shar-ma I made the changes as you asked.

N-Shar-ma reviewed (approved):
Not very well versed with this area of Zulip, but looks good to me!

alya commented:
Hm, I I get to an empty All messages, and then unmute a stream, I'm seeing errors popping up:

<img width="1170" alt="Screen Shot 2023-01-09 at 11 44 00 PM" src="https://user-images.githubusercontent.com/2090066/211491409-62a46fc7-b13e-41f9-8343-1e0509a0ad14.png">

SahilSingh177 commented:
> Hm, I I get to an empty All messages, and then unmute a stream, I'm seeing errors popping up:
> 
> <img alt="Screen Shot 2023-01-09 at 11 44 00 PM" width="1170" src="https://user-images.githubusercontent.com/2090066/211491409-62a46fc7-b13e-41f9-8343-1e0509a0ad14.png">

It is due to this line ``` message_lists.current.selected_row().offset() !== null``` in ```stream_muting.js```. I'll change it to ```message_lists.current.selected_row().offset() !== undefined``` to solve the error.

SahilSingh177 commented:
@alya I fixed the error.

SahilSingh177 commented:
Hey @alya, I am following up on this pull request and would appreciate if you could review it as it has been pending for some time now.

alya commented:
This worked for me when I first deleted the last message from All messages, but I'm not seeing the message anymore after I refresh the page.

<img width="913" alt="Screen Shot 2023-01-22 at 10 33 17 PM" src="https://user-images.githubusercontent.com/2090066/213978624-aacc2932-9812-4cd3-a32c-3c9debd51985.png">

Messages also fail to appear in the All messages view once I unmute a muted stream.

Please update the PR, and then test it *carefully*, trying out a variety of scenarios. Thanks!

SahilSingh177 commented:
@alya, I fixed the bug that you mentioned.
I performed manual testing by refreshing the page after clearing all messages and the results were satisfactory. Additionally, when a muted stream is unmuted, messages display in the All messages view.
I have updated the pull request description and included a video demonstration as reference.

alya commented:
@laurynmm Would you be up for reviewing this one?

laurynmm reviewed (commented):
@SahilSingh177 - Thanks for working on this issue!

I made a couple of comments on your current changes, but also noticed that there's still a bug when the user is in the all messages few for all public streams (there will be messages here that are not normally in the user's message history). This is the `/#narrow/streams/public` view.

With your updates and with a user that has all the topics in their message history muted / no private messages, if you go to the first message in that narrowed all public messages view and mute that first message's topic, then the empty narrow message will appear at the top of the screen even though there are still messages in the view. So looking into how the mute topic works for that narrow view and how it's interacting with these changes is the next part to figure out.

Also, just noting here that you should review the [guidelines for writing commit messages](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages). Both your summary and message body currently need to be updated.

Thanks again and let me know if you have any questions!

SahilSingh177 commented:
> @SahilSingh177 - Thanks for working on this issue!
> 
> I made a couple of comments on your current changes, but also noticed that there's still a bug when the user is in the all messages few for all public streams (there will be messages here that are not normally in the user's message history). This is the `/#narrow/streams/public` view.
> 
> With your updates and with a user that has all the topics in their message history muted / no private messages, if you go to the first message in that narrowed all public messages view and mute that first message's topic, then the empty narrow message will appear at the top of the screen even though there are still messages in the view. So looking into how the mute topic works for that narrow view and how it's interacting with these changes is the next part to figure out.
> 
> Also, just noting here that you should review the [guidelines for writing commit messages](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages). Both your summary and message body currently need to be updated.
> 
> Thanks again and let me know if you have any questions!

Through examination, I discovered that the origin of the bug was due to the UI not updating upon any changes in topics. To resolve the issue, I added the ```reset_ui_state()``` function to the ```handle_topic_updates()``` in the ```muted_topic_ui.js``` file, which effectively resolved the issue, as demonstrated in the attached video.

Previous behavior 

https://user-images.githubusercontent.com/96344003/215591268-3f54047d-85eb-40f9-b977-ff067ab8b329.mov

New behavior

https://user-images.githubusercontent.com/96344003/215591223-7ff3cab2-85e4-4052-8fdb-d610ca9591f2.mov

SahilSingh177 commented:
Hey @laurynmm, I am following up on this pull request and would appreciate if you could review it as it has been pending for some time now.

laurynmm commented:
@SahilSingh177 - Thanks for the ping! This PR got lost in my tracking as I was away for a bit the week before last.

Looking this over, I think that using `narrow.reset_ui_state` like this isn't going to work since it does a few other things besides check/set the empty narrow banner. 

For example, if you have the same user logged in to two browsers (or one browser in a regular and incognito session) and have one browser unread the messages in a topic and the other browser then mark that topic or stream as muted, then we start seeing some unexpected behaviours with the mark as unread banner disappearing and reappearing.

Another example, is having a search for a word open in one browser scrolled to the top so you see the banner noting that the results are from your history. Muting or unmuting a topic from that view will remove that banner notification from the top of the feed.

Here's a quick screencast of that second example with the changes from this PR...
<details>
<summary>Muting and search view example</summary>

[Screencast from 14-02-23 17:02:46.webm](https://user-images.githubusercontent.com/63245456/218791743-cd6fd51a-38be-4900-b085-047a70899589.webm)
</details>

So, we need to be able to update the empty narrow banner in these cases, but not these other parts of the UI, and still address the updating of the All messages view.

SahilSingh177 commented:
> @SahilSingh177 - Thanks for the ping! This PR got lost in my tracking as I was away for a bit the week before last.
> 
> Looking this over, I think that using `narrow.reset_ui_state` like this isn't going to work since it does a few other things besides check/set the empty narrow banner.
> 
> For example, if you have the same user logged in to two browsers (or one browser in a regular and incognito session) and have one browser unread the messages in a topic and the other browser then mark that topic or stream as muted, then we start seeing some unexpected behaviours with the mark as unread banner disappearing and reappearing.
> 
> Another example, is having a search for a word open in one browser scrolled to the top so you see the banner noting that the results are from your history. Muting or unmuting a topic from that view will remove that banner notification from the top of the feed.
> 
> Here's a quick screencast of that second example with the changes from this PR...
> 
> Muting and search view example
> So, we need to be able to update the empty narrow banner in these cases, but not these other parts of the UI, and still address the updating of the All messages view.

I agree, using ```narrow.reset_ui_state``` will not work as it is calling some functions that are causing bugs.

A possible way to fix this can be to create another function that only checks whether the narrow banner should be activated or not i.e, we should move the following code to a separate function:
```
    if (message_lists.current.empty()) {
        narrow_banner.show_empty_narrow_message();
    } else {
        narrow_banner.hide_empty_narrow_message();
    }

```
We can call this function instead of calling ```reset_ui_state()``` while muting streams/topics.
I tested the code for different scenarios and it seems to work perfectly.

Previous bug:

https://user-images.githubusercontent.com/96344003/219114077-8e52872b-bcfe-40d4-83b3-ef0fa28aeeb8.mov

Fix:

https://user-images.githubusercontent.com/96344003/219114135-0cbb5ad8-79dd-4a2c-94d5-f2719b92a9c7.mov

alya commented:
@laurynmm would be great to get another round of review from you once @SahilSingh177 has had a chance to resolve the conflicts.

SahilSingh177 commented:
@laurynmm, I have updated the PR with the fix discussed above [here](https://github.com/zulip/zulip/pull/23950#issuecomment-1431804422). Please review and let me know if you require any further changes. Thank you!

SahilSingh177 commented:
Hey @laurynmm, I am following up on this pull request and would appreciate if you could review it as it has been pending for some time now.

laurynmm commented on web/src/muted_topics_ui.js:
```diff
@@ -6,6 +6,7 @@ import * as channel from "./channel";
 import * as feedback_widget from "./feedback_widget";
 import {$t} from "./i18n";
 import * as message_lists from "./message_lists";
+import {handle_empty_narrow_banner} from "./narrow";
```
You should follow the pattern of importing the module vs just the function here.

laurynmm commented on web/src/narrow.js:
```diff
@@ -172,10 +172,19 @@ export function update_narrow_title(filter) {
     // Resets the state of various visual UI elements that are
     // a function of the current narrow.
-    narrow_banner.hide_empty_narrow_message();
+    handle_empty_narrow_banner();
```
This changes the behavior for narrows (a simple stream narrow is an example), which I don't think we want. The empty narrow banner appears before the web app has received the data from the server for the narrow.

You can see the different behavior between this and main if you use the dev tools in Chrome to throttle the network connection to 3G or Good 3G. In main, the empty narrow doesn't show up until the server response; you only see the logo with the spinner. With these changes, you see the empty narrow banner immediately while the logo with the spinner is waiting for the server response.

SahilSingh177 replied:
After following the steps you provided, I was able to obtain identical outcomes. 

Additionally, I observed that when we switch from a stream with messages to an empty one, the loading-spinner stops before the narrow banner is displayed. However, if we switch from one empty stream to another, the banner displays prior to the spinner coming to a halt.

https://user-images.githubusercontent.com/96344003/225755380-6f9652ac-acde-4539-9ce7-80b01b8a23ee.mov

laurynmm replied:
Yeah, the current behavior where the banner is removed and then reappears if/when no messages are received from the server should be maintained.

laurynmm reviewed (commented):
@SahilSingh177 - Thanks for the ping on this PR.

These changes still impact the current expected behavior of when the empty narrow banner is shown for a normal narrow view (e.g. a stream view). See my review note for how to reproduce / test.

I'm wondering if more work on #23812 (to make `message_list.home` not a special case) needs to happen before fixing this relatively rare case of an empty "All messages" view (most users will always have private messages from bots if no one else)?

Also, just noting here that you need to update your commit message body to for [these commit message guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#message-body).

Let me know if you have any questions about the above or review comments. Thanks!

SahilSingh177 commented:
@laurynmm, I have updated the PR with the following solution:
Added a function `handle_empty_narrow_message_home()` inside `narrow.js` that checks for messages in the 'All messages' screen and displays the narrow banner appropriately. 

I have tested the code for the following scenarios:

- Muting/unmuting a stream while inside it works as intended.
- Deleting the only message in "All messages" displays the narrow banner.
- Muting the only topic in "All messages" displays the narrow banner.
- Muting one stream does not affect others while inside it.
- The narrow banner only appears after the message has been fetched.

I have updated the PR description and added videos to demonstrate the changes.

SahilSingh177 commented:
Please review and let me know if any further changes are required. Thank you!

SahilSingh177 commented on web/src/message_list.js:
```diff
@@ -112,6 +113,7 @@ export class MessageList {
             render_info = this.append_to_view(bottom_messages, opts);
         }
 
+        narrow.handle_empty_narrow_message_home();
```
Calling the function here instead of removing `this.narrowed` to avoid issues such as the narrow banner not appearing when unmuting a topic within a stream.

SahilSingh177 commented on web/src/stream_muting.js:
```diff
@@ -23,6 +24,7 @@ export function update_is_muted(sub, value) {
         } else if (
             narrow_state.is_message_feed_visible() &&
             message_lists.current === message_lists.home &&
+            message_lists.current.selected_row().offset() !== undefined &&
```
I included this line of code because the console was throwing a "Cannot read properties of undefined (reading 'top')" error when attempting to unmute an empty stream on the "All Messages" screen.

SahilSingh177 commented:
Bumping this up for another round of review. FYI @laurynmm.
